### PR TITLE
HomeWork4.25.03.TrainingManualEsdStr.26

### PR DESCRIPTION
**Pull request by Java-Dev-Art**

### CODE REVIEW

File: src/HomeWork3/FirstTrainingManualToEnd/FirstTrainingManualToEnd.java [hunk 1]
```diff
@@ -324,9 +324,17 @@ public static void main(String[] args) {
 //                    twoDim[indexI][indexJ] = 1;
 //                }
 
+
```
Comment (svetamolchanova) [inline src/HomeWork3/FirstTrainingManualToEnd/FirstTrainingManualToEnd.java:327]: unnecessary spaces

Comment (svetamolchanova) [inline src/HomeWork3/FirstTrainingManualToEnd/FirstTrainingManualToEnd.java:327]: unnecessary spaces

File: src/HomeWork3/FirstTrainingManualToEnd/FirstTrainingManualToEnd.java [hunk 2]
```diff
@@ -324,9 +324,17 @@ public static void main(String[] args) {
...
+            }
+        }
+//        не бабочка  а крееест
+        for (int indexI = 0; indexI < twoDim.length; indexI++) {
```
Comment (svetamolchanova) [inline src/HomeWork3/FirstTrainingManualToEnd/FirstTrainingManualToEnd.java:331]: usually use a literal variable ("i", "j" ) for a counter variable

File: src/HomeWork3/FirstTrainingManualToEnd/FirstTrainingManualToEnd.java [hunk 3]
```diff
@@ -324,9 +324,17 @@ public static void main(String[] args) {
...
+            for (int indexJ = 0; indexJ < twoDim[indexI].length; indexJ++) {
+                System.out.print(twoDim[indexI][indexJ] + " ");
             }
+            System.out.println();
         }
-        System.out.println(Arrays.deepToString(twoDim));
+//        System.out.println(Arrays.deepToString(twoDim));
 //        int i, j;
 //
 //
```
Comment (svetamolchanova) [inline src/HomeWork3/FirstTrainingManualToEnd/FirstTrainingManualToEnd.java:340]: delete unnecessary comment lines

File: src/HomeWork4/HomeWork4.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        * Task # 18
+        * Найти в строке не только запятые, но и другие знаки препинания. Подсчитать общее их количество.
+        * */
+        System.out.println("Task #18");
+        String text = "Найти в строке не только запятые, но и другие знаки препинания. Подсчитать общее их количество.";
+        int index = 0;
+        char symbol;
+        for (int i = 0;i < text.length(); i++){
+            symbol = text.charAt(i);
+            if (symbol != 0){
```
Comment (svetamolchanova) [inline src/HomeWork4/HomeWork4.java]: what you check in this condition? symbol is one literal in string

File: src/HomeWork4/HomeWork4.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        * текста также могут быть пробелы, но могут и отсутствовать.
+        *
+        * */
+        System.out.println("Task #19");
+        String text19 = "Имеется строка с текстом. Подсчитать количество слов в тексте. Желательно\n" +
+                "учесть, что слова могут разделяться несколькими пробелами, в начале и конце\n" +
+                "текста также могут быть пробелы, но могут и отсутствовать.";
+        int countText19 = 0;
+        String [] strings = text19.split("\\s+");
+        for (int i = 0; i < strings.length - 1; i++){
```
Comment (svetamolchanova) [inline src/HomeWork4/HomeWork4.java]: why you check for < (strings.length - 1)? Try to enter array and see

File: src/HomeWork4/HomeWork4.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+            String text19C = strings[i];
+            if (text19C != null){
```
Comment (svetamolchanova) [inline src/HomeWork4/HomeWork4.java]: and for what we need this condition?

File: src/HomeWork4/HomeWork4.java [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        *
+        * */
+        System.out.println("Task #19");
+        String text19 = "Имеется строка с текстом. Подсчитать количество слов в тексте. Желательно\n" +
+                "учесть, что слова могут разделяться несколькими пробелами, в начале и конце\n" +
+                "текста также могут быть пробелы, но могут и отсутствовать.";
+        int countText19 = 0;
+        String [] strings = text19.split("\\s+");
+        for (int i = 0; i < strings.length - 1; i++){
+            String text19C = strings[i];
```
Comment (svetamolchanova) [inline src/HomeWork4/HomeWork4.java]: you can use strings[i]. Why we need new variable "text19C"?

File: src/HomeWork4/HomeWork4.java [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class HomeWork4 {
+    public static void main(String[] args) {
+        /*
+        * Task # 18
+        * Найти в строке не только запятые, но и другие знаки препинания. Подсчитать общее их количество.
+        * */
```
Comment (svetamolchanova) [inline src/HomeWork4/HomeWork4.java]: better each task to do in separate class

File: src/HomeWork4/HomeWork4.java [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+        System.out.println("Task #19");
+        String text19 = "Имеется строка с текстом. Подсчитать количество слов в тексте. Желательно\n" +
+                "учесть, что слова могут разделяться несколькими пробелами, в начале и конце\n" +
+                "текста также могут быть пробелы, но могут и отсутствовать.";
+        int countText19 = 0;
+        String [] strings = text19.split("\\s+");
+        for (int i = 0; i < strings.length - 1; i++){
+            String text19C = strings[i];
+            if (text19C != null){
+                countText19++;
```
Comment (svetamolchanova) [inline src/HomeWork4/HomeWork4.java]: also intesting moment. if we need get only amount elements we can get int countText = strings.length

File: src/HomeWork4/HomeWork4.java [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+        *
+        * */
+        System.out.println("Task #20");
+        String text20 ="Вывести текст, составленный из последних бук всех слов";
+        StringBuilder builder = new StringBuilder();
+        Pattern pattern = Pattern.compile("([а-яА-Я]+)");
+        Matcher matcher = pattern.matcher(text20);
+        int count20 = 0;
+        while (matcher.find()){
+            builder.append(text20.substring(matcher.end()-1,matcher.end()));
```
Comment (svetamolchanova) [inline src/HomeWork4/HomeWork4.java]: you can use another appand method:
append(CharSequence s, int start, int end)

File: src/HomeWork4/HomeWork4.java [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+        System.out.println("Task #21");
+        String text21 = "Напишите три цикла выполняющих многократное сложение строк, один с\n" +
+                "помощью оператора сложения и String, другой с помощью StringBuilder и метода\n" +
+                "append, а также аналогино для StringBuffer. Сравните скорость их выполнения.\n" +
+                "Регулярные выражения";
+        String[] text21Strings = text21.split("\\s+");
+        String text21S = "";
+        long start= System.nanoTime();
+
+
```
Comment (svetamolchanova) [inline src/HomeWork4/HomeWork4.java]: unnecessary spaces

File: src/HomeWork4/HomeWork4.java [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+        for (int i = 0; i< text21Strings.length; i++){
+            text21S += text21Strings[i];
+        }
+
+        long end =  System.nanoTime();
+        long result = end - start;
+
+        System.out.println("Добавление строк String . \nВремени затрачено на добавление :" + result / 1000 +
+                " мс. \nРезультат доьвыления : " + text21S );
```
Comment (svetamolchanova) [inline src/HomeWork4/HomeWork4.java]: all good, but it is not milliseconds. for example you can concatenate a string multiple times

File: src/HomeWork4/HomeWork4.java [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+        System.out.println("");
+
+        /*
+        *
+        * Task #22
+        * Написать программу, выполняющую поиск в строке шестнадцатеричных чисел,
+        * записаных по правилам Java, с помощью регулярных выражений и вывести их на страницу 3BD16.
+        *
+        *
+        *
```
Comment (svetamolchanova) [inline src/HomeWork4/HomeWork4.java]: unnecessary spaces

File: src/HomeWork4/HomeWork4.java [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+        * */
+        System.out.println("Task #22");
+        String task22 = "Написать программу, выполняющую поиск в строке 3A5 шестнадцатеричных чисел,\n" +
+                " * записаных по правилам Java, с помощью регулярных выражений и вывести их на страницу 3BD16.";
```
Comment (svetamolchanova) [inline src/HomeWork4/HomeWork4.java]: hexadecimal numbers start with "0x" or "0X"

File: src/HomeWork4/HomeWork4.java [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+        /*
+        * Task #23
+        * Написать программу, выполняющую поиск в строке всех тегов абзацев, в т.ч. тех, у
+        * которых есть параметры, например <p id=”p1”>, и замену их на простые теги абзацев <p>.
+        * */
+        System.out.println("Task #23");
+        String task23 = "Написать программу, выполняющую поиск в строке всех тегов абзацев, в т.ч. тех, у\n" +
+                "которых есть параметры, например <p id=”p1”>, и замену их на простые теги\n" +
+                "абзацев <p>.";
+        String tegTask23 = "<([a-z]\\s[a-z]{2}=.[a-z][0-9].)>";
```
Comment (svetamolchanova) [inline src/HomeWork4/HomeWork4.java]: if we have several parameters

File: src/HomeWork4/HomeWork4.java [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (svetamolchanova) [inline src/HomeWork4/HomeWork4.java]: unnecessary spaces

File: src/HomeWork4/HomeWork4.java (hunk 4 above)
Comment (Java-Dev-Art) [reply src/HomeWork4/HomeWork4.java]: I think. if we have some punctuation marks in text . Program find them and will count the number.

File: src/HomeWork4/Task22.java [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+         * Написать программу, выполняющую поиск в строке шестнадцатеричных чисел,
+         * записаных по правилам Java, с помощью регулярных выражений и вывести их на страницу 3BD16.
+         *
+         *
+         *
+         * */
+        System.out.println("Task #22");
+        String task22 = "Написать программу, выполняющую поиск в строке 3A5 шестнадцатеричных чисел,\n" +
+                " * записаных по правилам Java, с помощью регулярных выражений и вывести их на страницу c.";
+        Pattern patternTask22 = Pattern.compile("[0-9][a-zA-Z]+[0-9]+");
```
Comment (svetamolchanova) [inline src/HomeWork4/Task22.java:20]: Pattern is 0x... or 0X...

**Pull request closed without merging**